HPCC-16557 Default DESDL root paths on dali upon query

rpastrana opened the pull request:
- Create root DESDL paths on dali

Signed-off-by: Rodrigo Pastrana <rodrigo.pastrana@lexisnexis.com>

rpastrana committed:
HPCC-16557 Default DESDL root paths on dali upon query

- Create root DESDL paths on dali

Signed-off-by: Rodrigo Pastrana <rodrigo.pastrana@lexisnexis.com>

rpastrana commented:
@jakesmith please review

HPCCSmoketest commented:
Automated Smoketest
Sha: 03de3ddbec8b9ddd8ec01a18c872a11624332d3a
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 91.288 seconds |

Install hpccsystems-platform-community_6.2.0-rc5.el7.x86_64.rpm
HPCC Start: OK

Unittest result:

| total | passed | failed | timeout | 
|---|---|---|---|
|88 |88 |0 |0 |

HPCC Stop: OK
HPCC Uninstall: OK

jakesmith commented on esp/services/ws_esdlconfig/ws_esdlconfigservice.cpp:
```diff
@@ -1780,7 +1780,7 @@ bool CWsESDLConfigEx::onGetESDLDefinition(IEspContext &context, IEspGetESDLDefin
 bool CWsESDLConfigEx::onListESDLDefinitions(IEspContext &context, IEspListESDLDefinitionsRequest&req, IEspListESDLDefinitionsResponse &resp)
 {
-    Owned<IRemoteConnection> conn = querySDS().connect(ESDL_DEFS_ROOT_PATH, myProcessSession(), RTM_LOCK_READ, SDS_LOCK_TIMEOUT_DESDL);
+    Owned<IRemoteConnection> conn = querySDS().connect(ESDL_DEFS_ROOT_PATH, myProcessSession(), RTM_LOCK_READ|RTM_CREATE_QUERY, SDS_LOCK_TIMEOUT_DESDL);
```
As per previous comment:
```
Perhaps similar to previous comment, re. best not to overly use RTM_CREATE_QUERY where unnecessary. In this case the non-existence of the node is handled in the same way as it would be if a blank node was created via CREATE_QUERY, so it is prob. best not to add RTM_CREATE_QUERY here.
```

jakesmith commented:
@rpastrana - added a few comments, I don't _think_ you need the new RTM_CREATE_QUERY in most cases. If those methods only read and handle non-existence of node already, it is better if they just use the read lock they were using.

I can open a separate issue re. conn->close(false); conn->query* patterns, unless you want to comment/handler them in this PR ?

rpastrana commented on esp/services/esdl_svc_engine/esdl_binding.cpp:
```diff
@@ -105,7 +105,7 @@ IPropertyTree * fetchESDLBindingFromDali(const char *process, const char *bindin
+        conn.set(querySDS().connect(ESDL_BINDINGS_ROOT_PATH, myProcessSession(), RTM_LOCK_READ | RTM_CREATE_QUERY, SDS_LOCK_TIMEOUT_DESDL));
         if (!conn)
         {
             DBGLOG("Unable to connect to ESDL Service binding information in dali %s", ESDL_BINDINGS_ROOT_PATH);
```
@jakesmith I'm glad you commented on this, I believe I encountered issues w/ SDS locks and found that closing the connection immediately (even before querying the root) alleviated the lock contention, the scary part is that querying the root after the close always works. 

I will follow up via this issue: https://track.hpccsystems.com/browse/HPCC-16707

rpastrana commented on esp/services/ws_esdlconfig/ws_esdlconfigservice.cpp:
```diff
@@ -121,7 +121,7 @@ bool isESDLDefinitionBound(const char * esdldefid)
            return false;
 
-    Owned<IRemoteConnection> conn = querySDS().connect(ESDL_BINDINGS_ROOT_PATH, myProcessSession(), RTM_LOCK_READ, SDS_LOCK_TIMEOUT_DESDL);
+    Owned<IRemoteConnection> conn = querySDS().connect(ESDL_BINDINGS_ROOT_PATH, myProcessSession(), RTM_LOCK_READ | RTM_CREATE_QUERY, SDS_LOCK_TIMEOUT_DESDL);
```
Do you recommend making a single querySDS with RTM_CREATE_QUERY at loadtime?

rpastrana commented on esp/services/ws_esdlconfig/ws_esdlconfigservice.cpp:
```diff
@@ -121,7 +121,7 @@ bool isESDLDefinitionBound(const char * esdldefid)
+    Owned<IRemoteConnection> conn = querySDS().connect(ESDL_BINDINGS_ROOT_PATH, myProcessSession(), RTM_LOCK_READ | RTM_CREATE_QUERY, SDS_LOCK_TIMEOUT_DESDL);
     if (!conn)
        return false;
 
```
https://track.hpccsystems.com/browse/HPCC-16707

rpastrana commented on esp/services/ws_esdlconfig/ws_esdlconfigservice.cpp:
```diff
@@ -270,7 +270,7 @@ IPropertyTree * CWsESDLConfigEx::getESDLDefinitionRegistry(const char * wsEclId,
         throw MakeStringException(-1, "Unexpected error while attempting to access ESDL definition dali registry.");
 
-    Owned<IRemoteConnection> conn = querySDS().connect(ESDL_DEFS_ROOT_PATH, myProcessSession(), RTM_LOCK_WRITE, SDS_LOCK_TIMEOUT_DESDL);
+    Owned<IRemoteConnection> conn = querySDS().connect(ESDL_DEFS_ROOT_PATH, myProcessSession(), RTM_LOCK_WRITE|RTM_CREATE_QUERY, SDS_LOCK_TIMEOUT_DESDL);
```
I'm wondering if it might be worthwhile to ensure the root element is created (if it doesn't exist) at service load time...

rpastrana committed:
HPCC-16557 Ensure ESDL root paths at service load time

- Ensures ESDL/Bindings and ESDL/Definitions at WsESDLConfig load time
- Removes extraneous requests to create root elements on every wsesdlrequest

Signed-off-by: Rodrigo Pastrana <rodrigo.pastrana@lexisnexis.com>

rpastrana commented:
@jakesmith please review latest changes.

jakesmith commented on esp/services/esdl_svc_engine/esdl_binding.cpp:
```diff
@@ -36,6 +36,20 @@
+    if (!sdsPath)
+        return false;
+
+    Owned<IRemoteConnection> conn = querySDS().connect(sdsPath, myProcessSession(), RTM_CREATE_QUERY, SDS_LOCK_TIMEOUT_DESDL);
```
Should use a write lock explicitly really, i.e.: RTM_LOCK_WRITE|RTM_CREATE_QUERY

jakesmith commented on esp/services/esdl_svc_engine/esdl_binding.cpp:
```diff
@@ -36,6 +36,20 @@
+    Owned<IRemoteConnection> conn = querySDS().connect(sdsPath, myProcessSession(), RTM_CREATE_QUERY, SDS_LOCK_TIMEOUT_DESDL);
+    if (!conn)
+        return false;
+    conn->close(false);
```
You don't need this line, IRemoteConnection on destructor will do same.

HPCCSmoketest commented:
Automated Smoketest
Sha: d18da6cd1dcc33cb84a5468c310bebf6d8025d10
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 90.659 seconds |

Install hpccsystems-platform-community_6.2.0-rc5.el7.x86_64.rpm
HPCC Start: Fail
HPCC start:
Starting mydafilesrv ...      \x1b[32m [   OK    ]  \x1b[0m 
Starting mydali ...           \x1b[32m [   OK    ]  \x1b[0m 
Starting mydfuserver ...      \x1b[32m [   OK    ]  \x1b[0m 
Starting myeclagent ...       \x1b[32m [   OK    ]  \x1b[0m 
Starting myeclccserver ...    \x1b[32m [   OK    ]  \x1b[0m 
Starting myeclscheduler ...   \x1b[32m [   OK    ]  \x1b[0m 
Starting myesp ...            \x1b[31m [ FAILED  ]  \x1b[0m 
Starting myroxie ...          \x1b[32m [   OK    ]  \x1b[0m 
Starting mysasha ...          \x1b[32m [   OK    ]  \x1b[0m 
Starting mythor ...           \x1b[32m [   OK    ]  \x1b[0m 
HPCC status:
mydafilesrv     ( pid    10137 ) is running ...
mydali          ( pid    10292 ) is running ...
mydfuserver     ( pid    10461 ) is running ...
myeclagent      ( pid    10634 ) is running ...
myeclccserver   ( pid    10795 ) is running ...
myeclscheduler  ( pid    10963 ) is running ...
myesp           is stopped,
myroxie         ( pid    11308 ) is running ...
mysasha         ( pid    11497 ) is running ...
mythor          ( pid    11913 ) is running with 1 slave process(es) ...
myesp           is stopped,
pushd /mnt/disk1/home/vamosax/smoketest/PR-9374
~/smoketest/PR-9374 ~/smoketest/PR-9374/build
./unittest.sh
~/smoketest/PR-9374/build
HPCC Stop: OK
HPCC Uninstall: OK

jakesmith commented on esp/services/ws_esdlconfig/ws_esdlconfigservice.cpp:
```diff
@@ -398,10 +339,7 @@ void CWsESDLConfigEx::ensureESDLServiceBindingRegistry(const char * espProcName,
-        throw MakeStringException(-1, "Unable to connect to ESDL Service binding information in dali %s", ESDL_BINDINGS_ROOT_PATH);
-
-    Owned<IRemoteConnection> globalLock = querySDS().connect(ESDL_BINDINGS_ROOT_PATH, myProcessSession(), RTM_LOCK_WRITE, SDS_LOCK_TIMEOUT_DESDL);
+    Owned<IRemoteConnection> globalLock = querySDS().connect(ESDL_BINDINGS_ROOT_PATH, myProcessSession(), RTM_LOCK_WRITE | RTM_CREATE_QUERY, SDS_LOCK_TIMEOUT_DESDL);
```
Is this and similar elsewhere necessary any more? As you have:
```
    if(!ensureSDSPath(ESDL_BINDINGS_ROOT_PATH))
        throw MakeStringException(-1, "Could not ensure '%s' element on Dali", ESDL_BINDINGS_ROOT_PATH);
```
.. in CWsESDLConfigEx::init()

jakesmith commented on esp/services/ws_esdlconfig/ws_esdlconfigservice.cpp:
```diff
@@ -644,10 +582,7 @@ int CWsESDLConfigEx::publishESDLBinding(const char * bindingName,
-        throw MakeStringException(-1, "Unexpected error while attempting to access ESDL definition dali registry.");
-
-    Owned<IRemoteConnection> conn = querySDS().connect(ESDL_BINDINGS_ROOT_PATH, myProcessSession(), RTM_LOCK_WRITE, SDS_LOCK_TIMEOUT_DESDL);
+    Owned<IRemoteConnection> conn = querySDS().connect(ESDL_BINDINGS_ROOT_PATH, myProcessSession(), RTM_LOCK_WRITE | RTM_CREATE_QUERY, SDS_LOCK_TIMEOUT_DESDL);
```
See prev. comment

rpastrana replied:
So in this case, it comes down to how much overhead we're taking on due to the RTM_CREATE_QUERY, because there's always the chance that the /ESDL/Bindings tree in dali was removed by some other process... 

Publishing a new Binding shouldn't happen often, and we want to ensure the request completes even if the dali branch was deleted, so it seems worthwhile to keep the CREATE_QUERY.. thoughts?

jakesmith replied:
I think it's fine, was more curious why / when that it would expect to be missing if the branch had been ensured to exist at startup.

rpastrana replied:
it's not a common/supported case, but daliadmin actions could cause it... so... unless the extra cost is overwhelming, I think it's worthwhile.

jakesmith commented on esp/services/ws_esdlconfig/ws_esdlconfigservice.cpp:
```diff
@@ -1317,7 +1252,7 @@ IPropertyTree * CWsESDLConfigEx::getBindingTree(const char * espProcName, const
     }
 
-    Owned<IRemoteConnection> globalLock = querySDS().connect(ESDL_BINDINGS_ROOT_PATH, myProcessSession(), RTM_LOCK_WRITE|RTM_CREATE_QUERY, SDS_LOCK_TIMEOUT_DESDL);
+    Owned<IRemoteConnection> globalLock = querySDS().connect(ESDL_BINDINGS_ROOT_PATH, myProcessSession(), RTM_LOCK_WRITE, SDS_LOCK_TIMEOUT_DESDL);
```
See prev. comment

jakesmith commented:
@rpastrana - a few comments, I think you should add RTM_LOCK_WRITE to the ensure, the other comments may not be significant.

rpastrana committed:
HPCC-16557 Code review changes

- Removes unnecessary logic
- Ensures RTM_CREATE_QUERY requests are accompanied by RTM_LOCK_WRITE flag

Signed-off-by: Rodrigo Pastrana <rodrigo.pastrana@lexisnexis.com>

HPCCSmoketest commented:
Automated Smoketest
Sha: 4e8f8264c136208ae294266e78b91f042d9f44d2
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 90.648 seconds |

Install hpccsystems-platform-community_6.2.0-rc6.el7.x86_64.rpm
HPCC Start: Fail
HPCC start:
Starting mydafilesrv ...      \x1b[32m [   OK    ]  \x1b[0m 
Starting mydali ...           \x1b[32m [   OK    ]  \x1b[0m 
Starting mydfuserver ...      \x1b[32m [   OK    ]  \x1b[0m 
Starting myeclagent ...       \x1b[32m [   OK    ]  \x1b[0m 
Starting myeclccserver ...    \x1b[32m [   OK    ]  \x1b[0m 
Starting myeclscheduler ...   \x1b[32m [   OK    ]  \x1b[0m 
Starting myesp ...            \x1b[31m [ FAILED  ]  \x1b[0m 
Starting myroxie ...          \x1b[32m [   OK    ]  \x1b[0m 
Starting mysasha ...          \x1b[32m [   OK    ]  \x1b[0m 
Starting mythor ...           \x1b[32m [   OK    ]  \x1b[0m 
HPCC status:
mydafilesrv     ( pid    22883 ) is running ...
mydali          ( pid    23035 ) is running ...
mydfuserver     ( pid    23204 ) is running ...
myeclagent      ( pid    23376 ) is running ...
myeclccserver   ( pid    23537 ) is running ...
myeclscheduler  ( pid    23705 ) is running ...
myesp           is stopped,
myroxie         ( pid    24045 ) is running ...
mysasha         ( pid    24232 ) is running ...
mythor          ( pid    24651 ) is running with 1 slave process(es) ...
myesp           is stopped,
pushd /mnt/disk1/home/vamosax/smoketest/PR-9374
~/smoketest/PR-9374 ~/smoketest/PR-9374/build
./unittest.sh
~/smoketest/PR-9374/build
HPCC Stop: OK
HPCC Uninstall: OK

jakesmith commented:
@rpastrana - looks fine

@richardkchapman - good to merge.

richardkchapman commented:
Anyone know why the smoketest failed? @rpastrana

rpastrana commented:
@AttilaVamos how can we get a hold of the esp log for the smokebuild?

AttilaVamos commented:
@rpastrana Smoketest archives the content of  /var/log/HPCCSystems/ and I sent the latest to you via Flowdock.

AttilaVamos commented:
It seems there is a problem with couchbase:
``` 
"ERROR: 0: /mnt/disk1/home/vamosax/smoketest/PR-9374/HPCC-Platform/common/dllserver/thorplugin.cpp(588) : Loading plugin : Failed to load plugin /opt/HPCCSystems/plugins/libcouchbaseembed.so" 
```
There is not build error, but couple of warning for couchbaseembed.

rpastrana commented:
Thanks @AttilaVamos based on your findings, the esp issue might not be related to this particular PR... 
But I would like to see the entire log because the couchbase issue doesn't really make sense unless the couchbase pluging didn't build (would that cause a build time error?)

AttilaVamos commented:
The log collection sent to you via Flowdock and now resent by mail.

rpastrana commented:
The log I received seems to indicate an access issue to a function I moved in this latest commit. I'm not sure why

HPCCSmoketest commented:
Automated Smoketest
Sha: a694a78a4de44359f5fa82149970c04a99eaef4d
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 98.966 seconds |

Install hpccsystems-platform-community_6.2.0-rc6.el7.x86_64.rpm
HPCC Start: Fail
HPCC start:
Starting mydafilesrv ...      \x1b[32m [   OK    ]  \x1b[0m 
Starting mydali ...           \x1b[32m [   OK    ]  \x1b[0m 
Starting mydfuserver ...      \x1b[32m [   OK    ]  \x1b[0m 
Starting myeclagent ...       \x1b[32m [   OK    ]  \x1b[0m 
Starting myeclccserver ...    \x1b[32m [   OK    ]  \x1b[0m 
Starting myeclscheduler ...   \x1b[32m [   OK    ]  \x1b[0m 
Starting myesp ...            \x1b[31m [ FAILED  ]  \x1b[0m 
Starting myroxie ...          \x1b[32m [   OK    ]  \x1b[0m 
Starting mysasha ...          \x1b[32m [   OK    ]  \x1b[0m 
Starting mythor ...           \x1b[32m [   OK    ]  \x1b[0m 
HPCC status:
mydafilesrv     ( pid    10461 ) is running ...
mydali          ( pid    10614 ) is running ...
mydfuserver     ( pid    10783 ) is running ...
myeclagent      ( pid    10955 ) is running ...
myeclccserver   ( pid    11117 ) is running ...
myeclscheduler  ( pid    11288 ) is running ...
myesp           is stopped,
myroxie         ( pid    11630 ) is running ...
mysasha         ( pid    11818 ) is running ...
mythor          ( pid    12236 ) is running with 1 slave process(es) ...
myesp           is stopped,
pushd /mnt/disk1/home/vamosax/smoketest/PR-9374
~/smoketest/PR-9374 ~/smoketest/PR-9374/build
./unittest.sh
~/smoketest/PR-9374/build
HPCC Stop: OK
HPCC Uninstall: OK

rpastrana commented:
@wangkx please review

rpastrana committed:
HPCC-16557 Ensure SDS Path is referenced as a static function

Signed-off-by: Rodrigo Pastrana <rodrigo.pastrana@lexisnexis.com>

HPCCSmoketest commented:
Automated Smoketest
Sha: 4257d282314712c8d44d5c533796247108524893
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 91.675 seconds |

Install hpccsystems-platform-community_6.2.0-rc6.el7.x86_64.rpm
HPCC Start: OK

Unittest result:

| total | passed | failed | timeout | 
|---|---|---|---|
|88 |88 |0 |0 |

HPCC Stop: OK
HPCC Uninstall: OK

wangkx commented:
@rpastrana the last change looks fine.

rpastrana commented:
The issue was not showing up on Ubuntu, but the smoketest on Centos caught it.
@richardkchapman please merge. Thanks.
